core/src/test/java/org/elasticsearch/gateway/MetaDataWriteDataNodesTests.java

```diff
@@ -261,45 +124,71 @@ public void testMetaWrittenWhenIndexIsClosedAndMetaUpdated() throws Exception {
                 .endObject()
                 .endObject()).get();
 
-        GetMappingsResponse getMappingsResponse = client().admin().indices().prepareGetMappings("red_index").addTypes("doc").get();
-        assertNotNull(((LinkedHashMap) (getMappingsResponse.getMappings().get("red_index").get("doc").getSourceAsMap().get("properties"))).get("integer_field"));
-        // restart master with empty data folder and maybe red node
-        ((InternalTestCluster) cluster()).stopCurrentMasterNode();
-        masterNode = startMasterNode();
+        GetMappingsResponse getMappingsResponse = client().admin().indices().prepareGetMappings(index).addTypes("doc").get();
+        assertNotNull(((LinkedHashMap) (getMappingsResponse.getMappings().get(index).get("doc").getSourceAsMap().get("properties"))).get("integer_field"));
 
-        ensureGreen("red_index");
-        assertIndexInMetaState(redNode, "red_index");
-        assertIndexInMetaState(masterNode, "red_index");
-        clusterStateResponse = client().admin().cluster().prepareState().get();
-        assertThat(clusterStateResponse.getState().getMetaData().index("red_index").getState().name(), equalTo(IndexMetaData.State.CLOSE.name()));
-        getMappingsResponse = client().admin().indices().prepareGetMappings("red_index").addTypes("doc").get();
-        assertNotNull(((LinkedHashMap) (getMappingsResponse.getMappings().get("red_index").get("doc").getSourceAsMap().get("properties"))).get("integer_field"));
+        // make sure it was also written on red node although index is closed
+        awaitBusy(new Predicate<Object>() {
+            @Override
+            public boolean apply(Object input) {
+                try {
+                    ImmutableOpenMap<String, IndexMetaData> indicesMetaData = getIndicesMetaDataOnNode(dataNode);
+                    return ((LinkedHashMap) (indicesMetaData.get(index).getMappings().get("doc").getSourceAsMap().get("properties"))).get("integer_field") != null &&
+                            indicesMetaData.get(index).state().equals(IndexMetaData.State.CLOSE);
+                } catch (Exception e) {
+                    logger.info("caught exception while reading meta state: ", e);
+                    return false;
+                }
+            }
+        });
 
-    }
+        // try the same and see if this also works if node was just restarted
```

Comment:
Each node holds an array of indices it knows of and checks if it should write new meta data by looking up in this array. We need it because if an index is closed it will not appear in the shard routing and we therefore need to keep track of what we wrote before. However, when the node is restarted this array is empty and we have to fill it before we decide what we write. This is why I added an explicit test for it.